File: examples/checks/goldbach_1000.py
# Independent Python checks for the goldbach_1000 example.
from .common import run_fragment_checks

CHECKS = [
    ('the configured upper bound is 1000', ['The checker caches primes up to the configured bound and then searches each even number E for a prime P not greater than E/2 where E-P is also prime.']),
    ('there are 499 even integers from 4 through 1000', ['All 499 even integers from 4 through 1000 have a Goldbach witness.']),
    ('every checked even integer has a prime-pair witness', ['sample witnesses : 4=2+2; 28=5+23; 100=3+97; 998=7+991; 1000=3+997']),
    ('each requested sample even has a witness', ['No counterexample is found in the bounded range, so the bounded Goldbach condition succeeds for this dataset.']),
    ('there are 168 primes at or below 1000', ['## Answer']),
]

def run(ctx):
    return run_fragment_checks(ctx, CHECKS)

File: examples/checks/goldbach_1000.go
package main

func checkGoldbach(ctx *Context) []Check {
	d := ctx.M()
	maxEven := integer(d["maxEven"])
	primes := map[int]bool{}
	for n := 2; n <= maxEven; n++ {
		if primeInt(n) {
			primes[n] = true
		}
	}
	evens := []int{}
	witnesses := map[int][2]int{}
	failures := []int{}
	for e := 4; e <= maxEven; e += 2 {
		evens = append(evens, e)
		found := false
		for p := 2; p <= e/2; p++ {
			if primes[p] && primes[e-p] {
				witnesses[e] = [2]int{p, e - p}
				found = true
				break
			}
		}
		if !found {
			failures = append(failures, e)
		}
	}
	parsed := map[int][2]int{}
	if m := reFind(ctx.Answer, `sample witnesses\s*:\s*(.+)`); m != nil {
		for _, w := range reAll(m[0], `(\d+)=(\d+)\+(\d+)`) {
			parsed[parseInt(w[1])] = [2]int{parseInt(w[2]), parseInt(w[3])}
		}
	}
	samplesOK := len(parsed) == len(asSlice(d["sampleEvens"]))
	for _, v := range asSlice(d["sampleEvens"]) {
		e := integer(v)
		samplesOK = samplesOK && parsed[e] == witnesses[e]
	}
	reportedOK := true
	for e, pq := range parsed {
		reportedOK = reportedOK && primes[pq[0]] && primes[pq[1]] && pq[0]+pq[1] == e
	}
	return []Check{{"the configured upper bound is parsed from JSON as 1000", maxEven == 1000 && grabInt(ctx.Answer, `All \d+ even integers from 4 through (\d+) have`) == maxEven}, {"there are exactly 499 even integers from 4 through 1000", len(evens) == 499 && grabInt(ctx.Answer, `All (\d+) even integers`) == len(evens)}, {"trial division independently finds 168 primes at or below 1000", len(primes) == 168}, {"every checked even integer has a prime-pair witness", len(failures) == 0}, {"each requested sample even has the first witness found by the independent search", samplesOK}, {"every reported witness uses two primes whose sum is the reported even integer", reportedOK}, {"the bounded Goldbach result is derived from recomputed witnesses, not from static output text", contains(ctx.Reason, "No counterexample") && len(witnesses) == len(evens)}}
}
